File: Week 20/backtracking.cpp
/*
 * =============================================================================
 * Week 20 — Backtracking  (C++17)
 * =============================================================================
 *
 * Topics covered
 * --------------
 *   1. Permutations  (of distinct integers)
 *   2. Subsets  (power set)
 *   3. N-Queens
 *   4. Sudoku Solver
 *   5. Word Search  (2D grid)
 *   6. Generate Parentheses
 *
 * Complexity cheat-sheet
 * ----------------------
 *   permutations         O(n! * n)           |  Space O(n) stack + O(n! * n) output
 *   subsets              O(2^n * n)           |  Space O(n) stack + O(2^n * n) output
 *   n_queens             O(n!)               |  Space O(n)
 *   sudoku_solver        O(9^(empty cells))  |  Space O(81)
 *   word_search          O(m*n * 3^L)        |  Space O(L) stack
 *   generate_parens      O(4^n / sqrt(n))    |  Catalan number
 *
 * Build & run
 *   g++ -std=c++17 -O2 -o backtracking backtracking.cpp && ./backtracking
 * =============================================================================
 */

#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <cassert>
#include <sstream>
#include <functional>

// ---------------------------------------------------------------------------
// Helper
// ---------------------------------------------------------------------------
template <typename T>
std::string vec_str(const std::vector<T>& v) {
    std::ostringstream oss;
    oss << "[";
    for (std::size_t i = 0; i < v.size(); ++i) {
        if (i) oss << ", ";
        oss << v[i];
    }
    oss << "]";
    return oss.str();
}

template <typename T>
std::string vec2d_str(const std::vector<std::vector<T>>& v) {
    std::ostringstream oss;
    oss << "[\n";
    for (const auto& row : v) oss << "    " << vec_str(row) << "\n";
    oss << "  ]";
    return oss.str();
}

// ---------------------------------------------------------------------------
// 1. Permutations — generate all permutations of distinct integers
// ---------------------------------------------------------------------------
// Complexity:  Time O(n! * n)  |  Space O(n) recursion stack
std::vector<std::vector<int>> permutations(std::vector<int> nums) {
    std::vector<std::vector<int>> result;

    std::function<void(int)> backtrack = [&](int start) {
        if (start == static_cast<int>(nums.size())) {
            result.push_back(nums);
            return;
        }
        for (int i = start; i < static_cast<int>(nums.size()); ++i) {
            std::swap(nums[start], nums[i]);
            backtrack(start + 1);
            std::swap(nums[start], nums[i]);
        }
    };

    backtrack(0);
    return result;
}

// ---------------------------------------------------------------------------
// 2. Subsets — generate the power set
// ---------------------------------------------------------------------------
// Complexity:  Time O(2^n * n)  |  Space O(n) recursion stack
std::vector<std::vector<int>> subsets(const std::vector<int>& nums) {
    std::vector<std::vector<int>> result;
    std::vector<int> current;

    std::function<void(int)> backtrack = [&](int start) {
        result.push_back(current);
        for (int i = start; i < static_cast<int>(nums.size()); ++i) {
            current.push_back(nums[i]);
            backtrack(i + 1);
            current.pop_back();
        }
    };

    backtrack(0);
    return result;
}

// ---------------------------------------------------------------------------
// 3. N-Queens — place n queens on an n x n board
// ---------------------------------------------------------------------------
// Complexity:  Time O(n!)  |  Space O(n)
std::vector<std::vector<std::string>> n_queens(int n) {
    std::vector<std::vector<std::string>> solutions;
    std::vector<int> queens(n, -1);  // queens[row] = column

    std::vector<bool> cols(n, false);
    std::vector<bool> diag1(2 * n, false);  // row - col + n
    std::vector<bool> diag2(2 * n, false);  // row + col

    std::function<void(int)> solve = [&](int row) {
        if (row == n) {
            // Build board.
            std::vector<std::string> board(n, std::string(n, '.'));
            for (int r = 0; r < n; ++r) board[r][queens[r]] = 'Q';
            solutions.push_back(board);
            return;
        }
        for (int col = 0; col < n; ++col) {
            if (cols[col] || diag1[row - col + n] || diag2[row + col]) continue;

            queens[row] = col;
            cols[col] = diag1[row - col + n] = diag2[row + col] = true;
            solve(row + 1);
            cols[col] = diag1[row - col + n] = diag2[row + col] = false;
        }
    };

    solve(0);
    return solutions;
}

// ---------------------------------------------------------------------------
// 4. Sudoku Solver
// ---------------------------------------------------------------------------
// Complexity:  Time O(9^(empty cells)) worst case  |  Space O(81)
class SudokuSolver {
public:
    bool solve(std::vector<std::vector<int>>& board) {
        for (int r = 0; r < 9; ++r) {
            for (int c = 0; c < 9; ++c) {
                if (board[r][c] != 0) continue;

                for (int num = 1; num <= 9; ++num) {
                    if (is_valid(board, r, c, num)) {
                        board[r][c] = num;
                        if (solve(board)) return true;
                        board[r][c] = 0;
                    }
                }
                return false;  // no valid number => backtrack
            }
        }
        return true;  // all cells filled
    }

private:
    bool is_valid(const std::vector<std::vector<int>>& board, int row, int col, int num) {
        for (int i = 0; i < 9; ++i) {
            if (board[row][i] == num) return false;
            if (board[i][col] == num) return false;
        }
        int br = (row / 3) * 3, bc = (col / 3) * 3;
        for (int r = br; r < br + 3; ++r)
            for (int c = bc; c < bc + 3; ++c)
                if (board[r][c] == num) return false;
        return true;
    }
};

// ---------------------------------------------------------------------------
// 5. Word Search — find a word in a 2D grid
// ---------------------------------------------------------------------------
// Complexity:  Time O(m * n * 3^L)  |  Space O(L) recursion stack
bool word_search(std::vector<std::vector<char>>& board, const std::string& word) {
    int m = static_cast<int>(board.size());
    int n = static_cast<int>(board[0].size());
    constexpr int dx[] = {0, 0, 1, -1};
    constexpr int dy[] = {1, -1, 0, 0};

    std::function<bool(int, int, int)> dfs = [&](int r, int c, int idx) -> bool {
        if (idx == static_cast<int>(word.size())) return true;
        if (r < 0 || r >= m || c < 0 || c >= n) return false;
        if (board[r][c] != word[idx]) return false;

        char saved = board[r][c];
        board[r][c] = '#';  // mark visited

        for (int d = 0; d < 4; ++d) {
            if (dfs(r + dx[d], c + dy[d], idx + 1)) {
                board[r][c] = saved;
                return true;
            }
        }
        board[r][c] = saved;  // restore
        return false;
    };

    for (int r = 0; r < m; ++r)
        for (int c = 0; c < n; ++c)
            if (dfs(r, c, 0)) return true;
    return false;
}

// ---------------------------------------------------------------------------
// 6. Generate Parentheses — all valid combos of n pairs
// ---------------------------------------------------------------------------
// Complexity:  Time O(4^n / sqrt(n)) — nth Catalan number
std::vector<std::string> generate_parentheses(int n) {
    std::vector<std::string> result;
    std::string current;

    std::function<void(int, int)> backtrack = [&](int open, int close) {
        if (static_cast<int>(current.size()) == 2 * n) {
            result.push_back(current);
            return;
        }
        if (open < n) {
            current.push_back('(');
            backtrack(open + 1, close);
            current.pop_back();
        }
        if (close < open) {
            current.push_back(')');
            backtrack(open, close + 1);
            current.pop_back();
        }
    };

    backtrack(0, 0);
    return result;
}

// ---------------------------------------------------------------------------
// main — test cases
// ---------------------------------------------------------------------------
int main() {
    std::cout << "=== Week 20: Backtracking ===\n\n";

    // 1. Permutations
    {
        std::cout << "-- Permutations --\n";
        auto perms = permutations({1, 2, 3});
        assert(perms.size() == 6);
        std::cout << "  permutations({1,2,3}): " << perms.size() << " results\n";
        for (const auto& p : perms) std::cout << "    " << vec_str(p) << "\n";
        std::cout << "\n";
    }

    // 2. Subsets
    {
        std::cout << "-- Subsets --\n";
        auto subs = subsets({1, 2, 3});
        assert(subs.size() == 8);  // 2^3
        std::cout << "  subsets({1,2,3}): " << subs.size() << " results\n";
        for (const auto& s : subs) std::cout << "    " << vec_str(s) << "\n";
        std::cout << "\n";
    }

    // 3. N-Queens
    {
        std::cout << "-- N-Queens --\n";
        auto sols4 = n_queens(4);
        assert(sols4.size() == 2);
        std::cout << "  4-Queens: " << sols4.size() << " solutions\n";
        for (const auto& board : sols4) {
            for (const auto& row : board) std::cout << "    " << row << "\n";
            std::cout << "\n";
        }

        auto sols8 = n_queens(8);
        assert(sols8.size() == 92);
        std::cout << "  8-Queens: " << sols8.size() << " solutions\n\n";
    }

    // 4. Sudoku Solver
    {
        std::cout << "-- Sudoku Solver --\n";
        std::vector<std::vector<int>> board = {
            {5,3,0, 0,7,0, 0,0,0},
            {6,0,0, 1,9,5, 0,0,0},
            {0,9,8, 0,0,0, 0,6,0},

            {8,0,0, 0,6,0, 0,0,3},
            {4,0,0, 8,0,3, 0,0,1},
            {7,0,0, 0,2,0, 0,0,6},

            {0,6,0, 0,0,0, 2,8,0},
            {0,0,0, 4,1,9, 0,0,5},
            {0,0,0, 0,8,0, 0,7,9}
        };

        SudokuSolver solver;
        bool solved = solver.solve(board);
        assert(solved);
        assert(board[0][2] == 4);  // was 0
        std::cout << "  Solved:\n";
        for (const auto& row : board) {
            std::cout << "    ";
            for (int x : row) std::cout << x << " ";
            std::cout << "\n";
        }
        std::cout << "\n";
    }

    // 5. Word Search
    {
        std::cout << "-- Word Search --\n";
        std::vector<std::vector<char>> board = {
            {'A','B','C','E'},
            {'S','F','C','S'},
            {'A','D','E','E'}
        };
        assert(word_search(board, "ABCCED"));
        assert(!word_search(board, "ABCB"));
        std::cout << "  \"ABCCED\" found: true\n";
        std::cout << "  \"ABCB\" found:   false\n\n";
    }

    // 6. Generate Parentheses
    {
        std::cout << "-- Generate Parentheses --\n";
        auto parens = generate_parentheses(3);
        assert(parens.size() == 5);
        std::cout << "  n=3: " << parens.size() << " combinations\n";
        for (const auto& p : parens) std::cout << "    " << p << "\n";
    }

    std::cout << "\nAll Week 20 tests passed.\n";
    return 0;
}

File: Week 20/python/backtracking.py
"""
Week 20: Backtracking
======================
Backtracking is a systematic way to explore all possible configurations of a
search space.  It builds candidates incrementally and abandons ("backtracks")
a candidate as soon as it determines the candidate cannot lead to a valid
solution.

General template:
    def backtrack(state):
        if is_solution(state):
            record(state)
            return
        for choice in choices(state):
            make(choice)
            backtrack(state)
            undo(choice)        # <-- backtrack

Topics covered:
    1. Permutations
    2. Subsets (power set)
    3. Combinations (n choose k)
    4. N-Queens (return all solutions)
    5. Sudoku solver
    6. Word search in grid
    7. Generate valid parentheses
"""

from __future__ import annotations

from typing import List, Optional


# ---------------------------------------------------------------------------
# 1. Permutations
# ---------------------------------------------------------------------------
def permutations(nums: List[int]) -> List[List[int]]:
    """
    Generate all permutations of *nums*.

    Strategy: at each level, pick an unused element and recurse.

    Time:  O(n! * n)  — n! permutations, each taking O(n) to copy
    Space: O(n) recursion depth + O(n!) for results
    """
    result: List[List[int]] = []
    used = [False] * len(nums)
    current: List[int] = []

    def backtrack() -> None:
        if len(current) == len(nums):
            result.append(current[:])
            return
        for i in range(len(nums)):
            if not used[i]:
                used[i] = True
                current.append(nums[i])
                backtrack()
                current.pop()
                used[i] = False

    backtrack()
    return result


# ---------------------------------------------------------------------------
# 2. Subsets (Power Set)
# ---------------------------------------------------------------------------
def subsets(nums: List[int]) -> List[List[int]]:
    """
    Generate all subsets of *nums* (the power set).

    Strategy: for each element, decide to include it or not.
    Use an index to avoid duplicates.

    Time:  O(2^n * n)  — 2^n subsets, each up to O(n) to copy
    Space: O(n) recursion depth
    """
    result: List[List[int]] = []
    current: List[int] = []

    def backtrack(start: int) -> None:
        result.append(current[:])  # record every partial subset
        for i in range(start, len(nums)):
            current.append(nums[i])
            backtrack(i + 1)
            current.pop()

    backtrack(0)
    return result


# ---------------------------------------------------------------------------
# 3. Combinations (n choose k)
# ---------------------------------------------------------------------------
def combinations(n: int, k: int) -> List[List[int]]:
    """
    Generate all combinations of k numbers from 1..n.

    Strategy: similar to subsets but only record when we have exactly k elements.
    Prune: if remaining elements aren't enough to fill the combination, stop.

    Time:  O(C(n,k) * k)
    Space: O(k) recursion depth
    """
    result: List[List[int]] = []
    current: List[int] = []

    def backtrack(start: int) -> None:
        if len(current) == k:
            result.append(current[:])
            return
        # Pruning: need (k - len(current)) more elements, and only
        # (n - start + 1) are available.
        remaining_needed = k - len(current)
        for i in range(start, n - remaining_needed + 2):
            current.append(i)
            backtrack(i + 1)
            current.pop()

    backtrack(1)
    return result


# ---------------------------------------------------------------------------
# 4. N-Queens — return all solutions
# ---------------------------------------------------------------------------
def n_queens(n: int) -> List[List[str]]:
    """
    Place n queens on an n x n chessboard such that no two queens
    attack each other.  Return all distinct solutions.

    Each solution is a list of n strings, where 'Q' marks a queen and '.'
    marks an empty cell.

    Strategy: place queens row by row.  Track which columns and diagonals
    are under attack using sets.

    Time:  O(n!)  — upper bound; actual much less due to pruning
    Space: O(n) for the board state + O(n^2 * solutions) for results
    """
    result: List[List[str]] = []
    queens: List[int] = []  # queens[row] = column
    cols: set[int] = set()
    diag1: set[int] = set()  # row - col
    diag2: set[int] = set()  # row + col

    def backtrack(row: int) -> None:
        if row == n:
            # Build board representation
            board = []
            for r in range(n):
                line = ["."] * n
                line[queens[r]] = "Q"
                board.append("".join(line))
            result.append(board)
            return

        for col in range(n):
            if col in cols or (row - col) in diag1 or (row + col) in diag2:
                continue  # conflict — prune
            # Place queen
            queens.append(col)
            cols.add(col)
            diag1.add(row - col)
            diag2.add(row + col)

            backtrack(row + 1)

            # Remove queen (backtrack)
            queens.pop()
            cols.remove(col)
            diag1.remove(row - col)
            diag2.remove(row + col)

    backtrack(0)
    return result


# ---------------------------------------------------------------------------
# 5. Sudoku Solver
# ---------------------------------------------------------------------------
def solve_sudoku(board: List[List[str]]) -> bool:
    """
    Solve a 9x9 Sudoku puzzle in-place.

    Empty cells are represented by '.'.
    Returns True if the puzzle is solvable, False otherwise.

    Strategy:
        1. Find the next empty cell.
        2. Try digits '1'-'9'.
        3. If a digit is valid (no conflict in row, column, or 3x3 box),
           place it and recurse.
        4. If recursion fails, undo and try the next digit (backtrack).

    Time:  O(9^(empty_cells)) worst case, but pruning makes it much faster.
    Space: O(empty_cells) recursion depth
    """

    def is_valid(row: int, col: int, digit: str) -> bool:
        """Check if placing *digit* at (row, col) is valid."""
        # Check row
        if digit in board[row]:
            return False
        # Check column
        for r in range(9):
            if board[r][col] == digit:
                return False
        # Check 3x3 box
        box_r, box_c = 3 * (row // 3), 3 * (col // 3)
        for r in range(box_r, box_r + 3):
            for c in range(box_c, box_c + 3):
                if board[r][c] == digit:
                    return False
        return True

    def backtrack() -> bool:
        for r in range(9):
            for c in range(9):
                if board[r][c] == ".":
                    for digit in "123456789":
                        if is_valid(r, c, digit):
                            board[r][c] = digit
                            if backtrack():
                                return True
                            board[r][c] = "."  # undo
                    return False  # no valid digit — trigger backtrack
        return True  # all cells filled

    return backtrack()


# ---------------------------------------------------------------------------
# 6. Word Search in Grid
# ---------------------------------------------------------------------------
def word_search(board: List[List[str]], word: str) -> bool:
    """
    Determine if *word* exists in the *board* by following adjacent
    (horizontal/vertical) cells.  Each cell may be used at most once per path.

    Strategy: DFS/backtracking from every cell that matches word[0].

    Time:  O(m * n * 3^L)  where L = len(word).  From any cell we branch
           into at most 3 directions (we don't revisit the cell we came from).
    Space: O(L) recursion depth
    """
    if not board or not board[0] or not word:
        return False

    rows, cols = len(board), len(board[0])

    def backtrack(r: int, c: int, idx: int) -> bool:
        if idx == len(word):
            return True
        if r < 0 or r >= rows or c < 0 or c >= cols:
            return False
        if board[r][c] != word[idx]:
            return False

        # Mark cell as visited by temporarily replacing it.
        saved = board[r][c]
        board[r][c] = "#"

        for dr, dc in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
            if backtrack(r + dr, c + dc, idx + 1):
                board[r][c] = saved  # restore before returning
                return True

        board[r][c] = saved  # restore (backtrack)
        return False

    for r in range(rows):
        for c in range(cols):
            if backtrack(r, c, 0):
                return True
    return False


# ---------------------------------------------------------------------------
# 7. Generate Valid Parentheses
# ---------------------------------------------------------------------------
def generate_parentheses(n: int) -> List[str]:
    """
    Generate all combinations of *n* pairs of well-formed parentheses.

    Strategy: build the string character by character.
      - Add '(' if open_count < n.
      - Add ')' if close_count < open_count.

    Time:  O(4^n / sqrt(n))  — the n-th Catalan number
    Space: O(n) recursion depth
    """
    result: List[str] = []
    current: List[str] = []

    def backtrack(open_count: int, close_count: int) -> None:
        if open_count == n and close_count == n:
            result.append("".join(current))
            return
        if open_count < n:
            current.append("(")
            backtrack(open_count + 1, close_count)
            current.pop()
        if close_count < open_count:
            current.append(")")
            backtrack(open_count, close_count + 1)
            current.pop()

    backtrack(0, 0)
    return result


# ===========================================================================
# Test Cases
# ===========================================================================
if __name__ == "__main__":
    print("=" * 60)
    print("Week 20 — Backtracking")
    print("=" * 60)

    # --- Permutations ------------------------------------------------------
    print("\n--- Permutations ---")
    perms = permutations([1, 2, 3])
    print(f"permutations([1,2,3]): {len(perms)} results")
    assert len(perms) == 6
    assert [1, 2, 3] in perms and [3, 2, 1] in perms

    # --- Subsets -----------------------------------------------------------
    print("\n--- Subsets ---")
    subs = subsets([1, 2, 3])
    print(f"subsets([1,2,3]): {subs}")
    assert len(subs) == 8  # 2^3
    assert [] in subs and [1, 2, 3] in subs

    # --- Combinations ------------------------------------------------------
    print("\n--- Combinations ---")
    combs = combinations(4, 2)
    print(f"combinations(4, 2): {combs}")
    assert len(combs) == 6  # C(4,2)
    assert [1, 2] in combs and [3, 4] in combs

    # --- N-Queens ----------------------------------------------------------
    print("\n--- N-Queens ---")
    solutions_4 = n_queens(4)
    print(f"4-Queens: {len(solutions_4)} solutions")
    assert len(solutions_4) == 2
    for sol in solutions_4:
        for row in sol:
            print(f"  {row}")
        print()

    solutions_8 = n_queens(8)
    print(f"8-Queens: {len(solutions_8)} solutions")
    assert len(solutions_8) == 92

    # --- Sudoku Solver -----------------------------------------------------
    print("\n--- Sudoku Solver ---")
    board = [
        ["5", "3", ".", ".", "7", ".", ".", ".", "."],
        ["6", ".", ".", "1", "9", "5", ".", ".", "."],
        [".", "9", "8", ".", ".", ".", ".", "6", "."],
        ["8", ".", ".", ".", "6", ".", ".", ".", "3"],
        ["4", ".", ".", "8", ".", "3", ".", ".", "1"],
        ["7", ".", ".", ".", "2", ".", ".", ".", "6"],
        [".", "6", ".", ".", ".", ".", "2", "8", "."],
        [".", ".", ".", "4", "1", "9", ".", ".", "5"],
        [".", ".", ".", ".", "8", ".", ".", "7", "9"],
    ]
    assert solve_sudoku(board) is True
    print("Solved Sudoku:")
    for row in board:
        print(f"  {' '.join(row)}")
    # Verify no '.' remains
    assert all(cell != "." for row in board for cell in row)

    # --- Word Search -------------------------------------------------------
    print("\n--- Word Search ---")
    grid = [
        ["A", "B", "C", "E"],
        ["S", "F", "C", "S"],
        ["A", "D", "E", "E"],
    ]
    assert word_search(grid, "ABCCED") is True
    assert word_search(grid, "SEE") is True
    assert word_search(grid, "ABCB") is False
    print("word_search('ABCCED') = True")
    print("word_search('SEE')    = True")
    print("word_search('ABCB')   = False")

    # --- Generate Parentheses ----------------------------------------------
    print("\n--- Generate Valid Parentheses ---")
    parens = generate_parentheses(3)
    print(f"generate_parentheses(3): {parens}")
    assert len(parens) == 5
    expected = ["((()))", "(()())", "(())()", "()(())", "()()()"]
    assert sorted(parens) == sorted(expected)

    print("\nAll tests passed!")
